### DISCUSSION TITLE
zod trpc date validation throws error

### DISCUSSION THREAD

**Discussion by u007** [trpc/trpc#3080] [Q&A] [answered]
hi,
i tried to use zod on server side as input, but keep getting:
```
[ { "code": "invalid_type", "expected": "date", "received": "string", "path": [ "dateOfEvent" ], "message": "Expected date, received string" } ]
```
I can confirm the post call is sending data like this:
```
{
    "0": {
        "amount": 22,
        "dateOfEvent": "2022-10-10T16:00:00.000Z",
        "supportingDocs": []
    }
}
```
my react code:
```
const res = await client.claims.upload.mutate({
      amount: parseFloat(claimData.bill),
      dateOfEvent: dayjs(claimData.date).toDate(),
      supportingDocs: [],
    }
```

and my server is receiving with zod:
```
protectedProcedure.input(
    z.object({
      amount: z.number(),
      dateOfEvent: z.date(),
      supportingDocs: z.array(z.string()),
    })
  ).mutation(async({ input, ctx }) => {
///...
});
```

did I do anything wrong? typescript does not show any error

**Comment by KATT** [comment, marked as answer]
You need superjson transformer to send dates.

You can also use zod transform to send a numeric value or whatnot to represent the date and instead transform it on the server

**Reply by KATT** [reply]
See https://kitchen-sink.trpc.io/react-hook-form

**Reply by jacob-leger** [reply]
Thank you! You saved me 11+ hours of headache!

**Comment by u007** [comment]
ohh nice, will give it a try

**Comment by u007** [comment]
this works great https://trpc.io/docs/v10/data-transformers

im using my own context because I needed to wait for another context for authentication token.
i couldn't figure a way to wrap via withTRPC to use parent context value and also wait for context token to be set

so within my context, i have isReady to indicate when client is initiated with config that contains the token
